### PR TITLE
JDK-8303794: IEEEremainder problems are causing test failures after JDK-8302801

### PR DESCRIPTION
**Pull request by jddarcy**
Quick port of IEEEremainder; sorry for missing this earlier. Will add targeted tests in a follow-up fix.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303794](https://bugs.openjdk.org/browse/JDK-8303794): IEEEremainder problems are causing test failures after JDK-8302801 (**Bug** - P2) ⚠️ Issue is not open.

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12915/head:pull/12915` \
`$ git checkout pull/12915`

Update a local copy of the PR: \
`$ git checkout pull/12915` \
`$ git pull https://git.openjdk.org/jdk.git pull/12915/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12915`

View PR using the GUI difftool: \
`$ git pr show -t 12915`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12915.diff">https://git.openjdk.org/jdk/pull/12915.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12915#issuecomment-1459133557)

### CODE REVIEW

**Review by dholmes-ora — APPROVED** [review]
I can't attest to the correctness of the algorithm but I assume the necessary testing has been done in that regard. I can attest to this fixing the UnsatisfiedLinkError.

Thanks.

**Review by bplb — APPROVED** [review]

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 1]
```diff
@@ -3301,4 +3301,166 @@ static double compute(double x) {
...
+                }
+            }
+            hz=hx-hy;lz=lx-ly; if(Integer.compareUnsigned(lx, ly) < 0) hz -= 1;
+            if(hz>=0) {hx=hz;lx=lz;}
+
+            /* convert back to floating value and restore the sign */
+            if((hx|lx)==0)                  /* return sign(x)*0 */
+                return Zero[sx>>>31]; // unsigned shift
+            while(Integer.compareUnsigned(hx, 0x00100000) < 0) {          /* normalize x */
+                hx = hx+hx+(lx>>31); lx = lx+lx;
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/lang/FdLibm.java:3437]: Should be `>>>`

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 2]
```diff
@@ -3301,4 +3301,166 @@ static double compute(double x) {
...
+            lx = __LO(x);           /* low  word of x */
+            hy = __HI(y);           /* high word of y */
+            ly = __LO(y);           /* low  word of y */
+            sx = hx&0x80000000;             /* sign of x */
+            hx ^=sx;                /* |x| */
+            hy &= 0x7fffffff;       /* |y| */
+
+            /* purge off exception values */
+            if((hy|ly)==0||(hx>=0x7ff00000)||       /* y=0,or x not finite */
+               ((hy|((ly | -ly) >> 31))>0x7ff00000))     /* or y is NaN */
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/lang/FdLibm.java:3367]: Should be `>>>`

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 3]
```diff
@@ -3301,4 +3301,166 @@ static double compute(double x) {
...
+                    hx = hz+hz+(lz>>>31); lx = lz+lz; // unsigned shift
+                }
+            }
+            hz=hx-hy;lz=lx-ly; if(Integer.compareUnsigned(lx, ly) < 0) hz -= 1;
+            if(hz>=0) {hx=hz;lx=lz;}
+
+            /* convert back to floating value and restore the sign */
+            if((hx|lx)==0)                  /* return sign(x)*0 */
+                return Zero[sx>>>31]; // unsigned shift
+            while(Integer.compareUnsigned(hx, 0x00100000) < 0) {          /* normalize x */
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/lang/FdLibm.java:3436]: Should be `hx < 0x00100000`

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 4]
```diff
@@ -3301,4 +3301,166 @@ static double compute(double x) {
...
+            hx = __HI(x);           /* high word of x */
+            lx = __LO(x);           /* low  word of x */
+            hy = __HI(y);           /* high word of y */
+            ly = __LO(y);           /* low  word of y */
+            sx = hx&0x80000000;             /* sign of x */
+            hx ^=sx;                /* |x| */
+            hy &= 0x7fffffff;       /* |y| */
+
+            /* purge off exception values */
+            if((hy|ly)==0||(hx>=0x7ff00000)||       /* y=0,or x not finite */
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/FdLibm.java:3366]:
```suggestion
            if ((hy|ly)==0||(hx>=0x7ff00000)||       /* y=0,or x not finite */
```

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 5]
```diff
@@ -3301,4 +3301,166 @@ static double compute(double x) {
...
+               ((hy|((ly | -ly) >> 31))>0x7ff00000))     /* or y is NaN */
+                return (x*y)/(x*y);
+            if(hx <=hy) {
+                if((hx<hy)||(Integer.compareUnsigned(lx, ly) < 0)) return x;      /* |x|<|y| return x */
+                if(lx==ly)
+                    return Zero[sx>>>31];  /* |x|=|y| return x*0*/ // unsigned shift
+            }
+
+            /* determine ix = ilogb(x) */
+            if(hx<0x00100000) {     /* subnormal x */
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/FdLibm.java:3376]:
```suggestion
            if (hx<0x00100000) {     /* subnormal x */
```

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 6]
```diff
@@ -3301,4 +3301,166 @@ static double compute(double x) {
...
+                    for (iy = -1022,i=(hy<<11); i>0; i<<=1) iy -=1;
+                }
+            } else iy = (hy>>20)-1023;
+
+            /* set up {hx,lx}, {hy,ly} and align y to x */
+            if(ix >= -1022)
+                hx = 0x00100000|(0x000fffff&hx);
+            else {          /* subnormal x, shift x to normal */
+                n = -1022-ix;
+                if(n<=31) {
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/FdLibm.java:3398]:
```suggestion
                if (n<=31) {
```

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 7]
```diff
@@ -3301,4 +3301,166 @@ static double compute(double x) {
...
+                    hx = (hx<<n)|(lx>>>(32-n)); // unsigned shift
+                    lx <<= n;
+                } else {
+                    hx = lx<<(n-32);
+                    lx = 0;
+                }
+            }
+            if(iy >= -1022)
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/FdLibm.java:3406]:
```suggestion
            if (iy >= -1022)
```

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 8]
```diff
@@ -3301,4 +3301,166 @@ static double compute(double x) {
...
+                hy = 0x00100000|(0x000fffff&hy);
+            else {          /* subnormal y, shift y to normal */
+                n = -1022-iy;
+                if(n<=31) {
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/FdLibm.java:3410]:
```suggestion
                if (n<=31) {
```

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 9]
```diff
@@ -3301,4 +3301,166 @@ static double compute(double x) {
...
+                    ly <<= n;
+                } else {
+                    hy = ly<<(n-32);
+                    ly = 0;
+                }
+            }
+
+            /* fix point fmod */
+            n = ix - iy;
+            while(n-- != 0) {
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/FdLibm.java:3421]:
```suggestion
            while (n-- != 0) {
```

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 10]
```diff
@@ -3301,4 +3301,166 @@ static double compute(double x) {
...
+                hz=hx-hy;lz=lx-ly; if(Integer.compareUnsigned(lx, ly) < 0 ) hz -= 1;
+                if(hz<0){hx = hx+hx+(lx>>>31); lx = lx+lx;} // unsigned shift
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/FdLibm.java:3423]:
```suggestion
                if (hz<0){hx = hx+hx+(lx>>>31); lx = lx+lx;} // unsigned shift
```

File: src/java.base/share/classes/java/lang/FdLibm.java [hunk 11]
```diff
@@ -3301,4 +3301,166 @@ static double compute(double x) {
...
+                else {
+                    if((hz|lz)==0)          /* return sign(x)*0 */
```
Comment (turbanoff) [inline src/java.base/share/classes/java/lang/FdLibm.java:3425]:
```suggestion
                    if ((hz|lz)==0)          /* return sign(x)*0 */
```

**Pull request closed without merging**